OrbitVulkanLayer/VulkanLayerController.h

```diff
@@ -0,0 +1,454 @@
+    if (result != VK_SUCCESS) {
+      return result;
+    }
+
+    std::vector<VkExtensionProperties> extensions(num_other_extensions);
+    result = dispatch_table_.EnumerateDeviceExtensionProperties(physical_device)(
+        physical_device, nullptr, &num_other_extensions, extensions.data());
+    if (result != VK_SUCCESS) {
+      return result;
+    }
+
+    // Lets append all of our extensions, that are not yet listed.
+    // Note, as this list of our extensions is very small, we are fine with O(N*M) runtime.
+    for (const auto& extension : kDeviceExtensions) {
+      bool is_unique = true;
+      for (const auto& other_extension : extensions) {
+        if (strcmp(extension.extensionName, other_extension.extensionName) == 0) {
+          is_unique = false;
+          break;
+        }
+      }
+      if (is_unique) {
+        extensions.push_back(extension);
+      }
+    }
+
+    // As above, if properties is nullptr, only the number if extensions is queried.
+    if (properties == nullptr) {
+      *property_count = extensions.size();
+      return VK_SUCCESS;
+    }
+
+    uint32_t num_extensions_to_copy = extensions.size();
+    // In the case that less extensions are queried then the layer uses, we copy on this number and
+    // return VK_INCOMPLETE, according to the specification.
+    if (*property_count < num_extensions_to_copy) {
+      num_extensions_to_copy = *property_count;
+    }
+    memcpy(properties, extensions.data(), num_extensions_to_copy * sizeof(VkExtensionProperties));
+    *property_count = num_extensions_to_copy;
+
+    if (num_extensions_to_copy < extensions.size()) {
+      return VK_INCOMPLETE;
+    }
+    return VK_SUCCESS;
+  }
+
+  [[nodiscard]] const DispatchTable* dispatch_table() const { return &dispatch_table_; }
+
+  [[nodiscard]] const SubmissionTracker* submission_tracker() const { return &submission_tracker_; }
+
+  [[nodiscard]] const DeviceManager* device_manager() const { return &device_manager_; }
+
+  [[nodiscard]] const TimerQueryPool* timer_query_pool() const { return &timer_query_pool_; }
+
+  [[nodiscard]] const QueueManager* queue_manager() const { return &queue_manager_; }
+
+ private:
+  void InitVulkanLayerProducerIfNecessary() {
+    absl::MutexLock lock{&vulkan_layer_producer_mutex_};
+    if (vulkan_layer_producer_ == nullptr) {
+      vulkan_layer_producer_ = std::make_unique<VulkanLayerProducerImpl>();
+      // TODO: As soon as we do have producer side channel for the port, use it.
+      vulkan_layer_producer_->BringUp(nullptr);
+      submission_tracker_.SetVulkanLayerProducer(vulkan_layer_producer_.get());
+    }
+  }
+
+  void CloseVulkanLayerProducerIfNecessary() {
+    absl::MutexLock lock{&vulkan_layer_producer_mutex_};
+    if (vulkan_layer_producer_ != nullptr) {
+      // TODO: Only do this when DestroyInstance has been called a number of times
+      //  equal to the number of times CreateInstance was called.
```

Comment:
Forgot about this... Should we address this TODO?

Reply:
We can also do this next year, I guess.

Reply:
Fine by me. I guess we should address it together with the other TODO above.

Reply:
Yep, lets do this then next year.